### PR TITLE
[hold pub] added function to calculate angle between base planes

### LINKED ISSUE #3478: angle between the planes of two nucleobases
## Is your feature request related to a problem? ##

I wanted to analyze the nucleic acids for the angle between the base planes and could not find anything simple and quick.

## Describe the solution you'd like ##

It would be great to have a function that takes in the resids for the two bases and returns the angle between the planes of the bases.

Adding it to the analysis module for nucleic acids sounds like best option.

### PR DESCRIPTION
**Pull request by aakognole**
Fixes #3478 

Changes made in this Pull Request:
 - added function to calculate angle between base planes

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/nuclinfo.py]: I'm not sure I agree with the premise of having hardcoded atom names and residues for things like this. It is always best for things like this to leave the selection to users and instead ask that folks pass in either directly coordinates or atomgroups.

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/nuclinfo.py:774]: Please do attend to the various PEP8 issues highlighted by pep8speaks https://github.com/MDAnalysis/mdanalysis/pull/3480#issuecomment-989321451

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/nuclinfo.py]:
```suggestion
    .. versionadded:: 2.1.0
```

**Review by IAlibay — CHANGES_REQUESTED** [review]
Hi @aakognole,

Thanks for opening this PR.

This is not the type of system I generally deal with and it's hard for me to conceptualise just how common an operation this is, and therefore a) if it should be included in the core library, and b) where it should be included.

If this is a common analysis, then this may be better implemented as an [AnalysisBase](https://docs.mdanalysis.org/2.0.0/documentation_pages/analysis/base.html?highlight=analysisbase) class rather than a function. If it's more of a method that can be generalised across various systems then maybe it should go into mdamath instead.

I am hoping one of the other @MDAnalysis/coredevs (maybe @orbeckst) has a better idea here?

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/nuclinfo.py]: Admittedly, this follows the style of the (very old) nuclinfo module. It would be consistent with the legacy code that's already there.

**Comment by orbeckst** [comment]
@aakognole do you have one or two paper citations where this analysis was used and/or described?

**Comment by aakognole** [comment]
Hi @orbeckst 

Yes, the angle between the base planes can be important criteria in determining weather the bases are considered stacked or not. I am working on something where I need to consider this as a colvar.

Here are a few references
https://doi.org/10.1016/0263-7855(95)00086-0
https://doi.org/10.1021/jp209986y

File: testsuite/MDAnalysisTests/analysis/test_nuclinfo.py
Comment (tylerjereddy) [inline testsuite/MDAnalysisTests/analysis/test_nuclinfo.py]: maybe `assert_allclose()` for new code: https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_almost_equal.html

Comment (aakognole) [reply testsuite/MDAnalysisTests/analysis/test_nuclinfo.py]: Okay. Would you like me to move to `assert_allclose()` for all the tests in nuclinfo? 

Although I think it is reasonable to stick with `assert_almost_equal()` since we are not looking for precision beyond 3 decimals.

Comment (tylerjereddy) [reply testsuite/MDAnalysisTests/analysis/test_nuclinfo.py]: No, just here for now; I can't think of a reason to use the discouraged approach in new code.

Comment (aakognole) [reply testsuite/MDAnalysisTests/analysis/test_nuclinfo.py]: Updated. Thanks.

**Comment by orbeckst** [comment]
Is the *Gabb et al, Journal of MolecularGraphics14:6-1 I. 1996* paper the original paper where they described the base plane angle the first time (tau_ij in the paper) ?

If so, please include the reference in the documentation, the [userguide has an entry on our doc style](https://userguide.mdanalysis.org/stable/contributing_code.html#working-with-the-code-documentation), in particular the [numpy doc string standard](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard). Add the citation itself to [doc/sphinx/source/documentation_pages/references.rst](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/doc/sphinx/source/documentation_pages/references.rst) instead of a references section in the function docs.

As an example look at [AtomGroup.asphericity](https://docs.mdanalysis.org/stable/documentation_pages/core/groups.html#MDAnalysis.core.groups.AtomGroup.asphericity).

**Comment by aakognole** [comment]
Hi @orbeckst 

> Is the _Gabb et al, Journal of MolecularGraphics14:6-1 I. 1996_ paper the original paper where they described the base plane angle the first time (tau_ij in the paper) ?
> 
Yes, this is the reference I could trace down for base plane angle analysis used for stacking. However the angle_between_base_planes in the paper is _theta_ij_ and not _tau_ij_. The _tau_ij_ refers to a different angle not used frequently. I would also refer to Jafilan et. al. paper since it has developed on this analysis and gives simpler stacking coordinate based on the angle_between_base_planes.

> If so, please include the reference in the documentation, the [userguide has an entry on our doc style](https://userguide.mdanalysis.org/stable/contributing_code.html#working-with-the-code-documentation), in particular the [numpy doc string standard](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard). Add the citation itself to [doc/sphinx/source/documentation_pages/references.rst](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/doc/sphinx/source/documentation_pages/references.rst) instead of a references section in the function docs.
> 
> As an example look at [AtomGroup.asphericity](https://docs.mdanalysis.org/stable/documentation_pages/core/groups.html#MDAnalysis.core.groups.AtomGroup.asphericity).

Thanks for the guidelines. I will add the references accordingly.

**Comment by orbeckst** [comment]
Very good, ping me when you need me to look at it.

**Comment by aakognole** [comment]
@orbeckst please take a look at the changes. Thanks.

**Review by orbeckst — CHANGES_REQUESTED** [review]
Please integrate the new functions in the docs. You need to mention them explicitly at the top of the file. Possibly add a new section under which the new functions fits if none of the other sections work.

Check that the docs are visible in the RTD build https://mdanalysis--3480.org.readthedocs.build/en/3480/documentation_pages/analysis/nuclinfo.html

**Review by orbeckst — APPROVED** [review]
Looks good. Thank you @aakognole – sorry it took such a long time.

**Comment by orbeckst** [comment]
@IAlibay do you want to have a final look and/or wait for Azure to run again (in which case, please kick it once the brownout is over). Cheers!

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/nuclinfo.py]: I realise this is in keeping with the other functions in this module, but segments is kind of a weird one because not everything really fully supports it - given the recent discussions on "chainID" vs "segID" would it make sense to switch from being resid/system for each base to just "a selection string for that base"? Would that be easier?

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/nuclinfo.py:850]: even if we disregard my comment above, I believe resids are not guaranteedto be unique (I always forget - am I right here @lilyminium ?) Would it make sense to do a check for > 1 residues in this atomgroup and throw a warning that we're taking the first residue if that's the case?

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/nuclinfo.py]: This will throw a KeyError if the residue is not part of batms and an IndexError if the names are wrong (since you're doing a positions call on a zero-width AtomGroup).

In both cases this will be quite incomprehensible to users if it fails. Could this be reformatted to add some try/excepts to make it clear to users as to why things fail (if they do)?

**Review by IAlibay — CHANGES_REQUESTED** [review]
Couple of comments here - please feel free to disregard my comment re: switching away from using segID (pinging @orbeckst) if it's too much of a change.

I would definitely like to see a bit more usage safety checking on the selections though.

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (aakognole) [reply package/MDAnalysis/analysis/nuclinfo.py:850]: @IAlibay Thanks for your comments. I believe a combination of segid and resid should point to a unique residue in case of polynucleotides.

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (aakognole) [reply package/MDAnalysis/analysis/nuclinfo.py]: In this case we can only have the 5 residues that make up the DNA and RNA systems. We already inform the users that this analysis module is based on CHARMM specific names for nucleic acid residues and atom names. Again, the user just needs to check that the segid and resid are correct.

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (IAlibay) [reply package/MDAnalysis/analysis/nuclinfo.py:850]: I am willing to defer to @orbeckst here, but personally I really am not sure that this is the case for all topology types we support. Segments are very charmm-centric, and `resid` is non-unique. If we use the PDB format as an example, wouldn't this easily trip up if you had an altloc for a residue with resid N?

Given the number of `select_atoms` calls we're doing here, performance is really not going to be limited to a simple check on `len(resn1.residues)`.

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/nuclinfo.py]: Normally I'd agree with you. However, for historical reasons all other functions do the same. In the interest of a common and consistent API I'd keep it and rather open an issue to rework the whole thing.

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (IAlibay) [reply package/MDAnalysis/analysis/nuclinfo.py]: This is putting a lot of onus on users to follow our docs and fully understand everything we mean. Personally, given that wrapping in a try/except here has nearly no cost and will immensely clarify failures, I fail to see how this is not worth it.

Again happy to defer to the opinions of @orbeckst as the original approving reviewer here.

Comment (orbeckst) [reply package/MDAnalysis/analysis/nuclinfo.py]: I agree with @IAlibay here — even though it's the user's job to put in the right selection, telling them if they don't is a better approach than just failing with unintelligible stack traces (because then _we_ have to explain on the mailing list what the error means). 

The other functions don't have these additional checks but your function could do better than the old ones. (And this is not an API change but usability improvement.)

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (IAlibay) [reply package/MDAnalysis/analysis/nuclinfo.py]: That's fair, I'm happy to go with that philosophy.

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/nuclinfo.py:850]: Valid argument, @IAlibay .

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (aakognole) [reply package/MDAnalysis/analysis/nuclinfo.py]: Please take a look. Thanks!

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/nuclinfo.py:28]: Only just picked this up - @orbeckst I've never been super clear on how these authorship entries worked in MDA, is this the correct way to address this? (and also do we have a clear view on when we add someone and when we don't?)

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/nuclinfo.py]: I think this is incorrect syntax (i.e. adding a `format` in the middle of a string continuation).

If anything this needs to be fixed - probably using f-strings here would be better?

Aside from that your construct isn't wrong but I think we can clear this up a little bit, I'll let you make the choice if you want to squash it down to fewer lines with the following loop:

```suggestion
    for ag, segid, res in [(bf1, seg1, b1), (bf2, seg2, b2)]:
        if len(ag.residues) > 1:
            warnings.warn(f"Found more than one residues with resid {res} and "
                          f"segid {segid}. Using first residue in selection.")
```

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/nuclinfo.py]: So the approach we would take here would be to raise the message with the error. Print calls can often be lost amongst a stack trace.

```suggestion
        errmsg = ("Residue names may be incorrect! "
                  "Please use either ADE/GUA/CYT/URA/THY or A/G/C/U/T format.")
        raise KeyError(errmsg)
```

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/nuclinfo.py]: Similarly to the above:

```suggestion
        errmsg = ("Atom names may be incorrect! "
                  "Make sure base atom names follow CHARMM format.")
        raise IndexError(errmsg)
```

File: testsuite/MDAnalysisTests/analysis/test_nuclinfo.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_nuclinfo.py:191]: Could you add in tests for the warnings being raised and the two exceptions please?

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/nuclinfo.py:870]: It's quite rare to have two `except` clauses, but I think it's justified here.

**Review by IAlibay — CHANGES_REQUESTED** [review]

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/nuclinfo.py:28]: No real guide lines and I would be liberal with it for right now.

Truth be told, I'd actually like to get rid of these entries (which come from the earliest days...) to make the docs more uniform and because AUTHORS is really the most appropriate part for noting contributions as it makes clear that this is a collective work.

So, I'd just leave it as is, with the understanding that all of these entries will maybe suddenly disappear .

Comment (IAlibay) [reply package/MDAnalysis/analysis/nuclinfo.py:28]: Yeah sounds good.

> Truth be told, I'd actually like to get rid of these entries (which come from the earliest days...) to make the docs more uniform and because AUTHORS is really the most appropriate part for noting contributions as it makes clear that this is a collective work.

That's also the feeling I have (hence the query here - I've just never reviewed a case of someone adding to the notice so I've just ignored them until now). That being said, we'd probably have to poll the individuals that wrote these notices to begin with, I'll raise an issue.

**Comment by fiona-naughton** [comment]
I had a quick look at the [Gabb et al. paper](https://www.sciencedirect.com/science/article/abs/pii/0263785595000860), and it looks like you're defining the base pane in a different way (using atoms N1/N3/C5 or N9/N1/N7 rather than geometric center+two atoms). If we're assuming planarity of the rings it shouldn't affect the result, but is there are particular reason for you choice (is there a standard way it's usually done?). Might be worth explicitly mentioning how you calculate it here, since it differs from the cited paper.

(Also, the second paper ([Jafilin et al](https://pubs.acs.org/doi/full/10.1021/jp209986y)), seems to be about developing a stacking coordinate that uses the angle-between-base-planes rather than about the angle-between-base-planes calculation itself - including it with please-cite here seems unnecessary to me? )

**Comment by aakognole** [comment]
> I had a quick look at the [Gabb et al. paper](https://www.sciencedirect.com/science/article/abs/pii/0263785595000860), and it looks like you're defining the base pane in a different way (using atoms N1/N3/C5 or N9/N1/N7 rather than geometric center+two atoms). If we're assuming planarity of the rings it shouldn't affect the result, but is there are particular reason for you choice (is there a standard way it's usually done?). Might be worth explicitly mentioning how you calculate it here, since it differs from the cited paper.

You are right, it is not exactly the same way. Assuming the planarity of the rings and using a modified selection of atoms to be consistent across the bases gives correct direction of the normal to the base plane. Mentioning the difference is a good idea. Thanks!

> (Also, the second paper ([Jafilin et al](https://pubs.acs.org/doi/full/10.1021/jp209986y)), seems to be about developing a stacking coordinate that uses the angle-between-base-planes rather than about the angle-between-base-planes calculation itself - including it with please-cite here seems unnecessary to me? )

Okay, yes although it is a good example of how the angle is used in defining the stacking coordinate, it may not be necessary to cite it.

**Comment by IAlibay** [comment]
> You are right, it is not exactly the same way. Assuming the planarity of the rings and using a modified selection of atoms to be consistent across the bases gives correct direction of the normal to the base plane. Mentioning the difference is a good idea. Thanks!

Is this used/validated anywhere else? Ideally we would want to justify to users why we're taking a shortcut here. As far as I can tell doing either the Cremer Pople based approach or the geometric center + 2 atoms isn't necessarily that much more complicated than what is done here, so it would be good to justify reasoning.

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (richardjgowers) [inline package/MDAnalysis/analysis/nuclinfo.py]: This isn't PBC aware and probably should be. See minimize_vectors which recently got added to lib.

**Review by richardjgowers — CHANGES_REQUESTED** [review]

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (aakognole) [reply package/MDAnalysis/analysis/nuclinfo.py]: good point. Usually users recenter the trajectories before analysis and this may not be always necessary. So do you think in this case the box input to minimize_vectors should be optional or not?

File: package/MDAnalysis/analysis/nuclinfo.py
Comment (richardjgowers) [inline package/MDAnalysis/analysis/nuclinfo.py:877]: Don't invent a box if there isn't one, it's incredibly hard to guess a box. Better to only do the call to minimize if there's a box, otherwise trust the vectors to be correct.

**Review by richardjgowers — CHANGES_REQUESTED** [review]

File: testsuite/MDAnalysisTests/analysis/test_nuclinfo.py
Comment (aakognole) [reply testsuite/MDAnalysisTests/analysis/test_nuclinfo.py:191]: Added.

**Comment by orbeckst** [comment]
@aakognole from the discussion above https://github.com/MDAnalysis/mdanalysis/pull/3480#issuecomment-1019553272 I gather that the method you propose here is different from the one in cited references. Is the approach here your own invention then? 

Did you use it in one of your published papers or are you about to publish a paper with it?

**Comment by aakognole** [comment]
> @aakognole from the discussion above [#3480 (comment)](https://github.com/MDAnalysis/mdanalysis/pull/3480#issuecomment-1019553272) I gather that the method you propose here is different from the one in cited references. Is the approach here your own invention then?

Hi @orbeckst . There is hardly any invention here. It is simpler yet effective approximation of finding the normal to the base planes and getting angle between them to use further. 

> Did you use it in one of your published papers or are you about to publish a paper with it?

Not published yet, but I am working on it. You may put merging this PR on hold until then. Thanks.

**Comment by IAlibay** [comment]
> Not published yet, but I am working on it. You may put merging this PR on hold until then. Thanks.

Just to keep you updated on this @aakognole . We are currently having some internal discussions amongst the core developer team on how to go forward with cases like this in the future.

Essentially the consensus is that we prefer not adding experimental (especially pre-publication) methods to the core library. Our aim in the near future will be to encourage folks to share new methods via an MDAKit (i.e. a self-contained python package that lives downstream from the core library).

In this particular case we are in agreement that we would at least like to wait until this is properly validated (i.e. published) _before_ we make a decision on merging. To this end I have added [hold pub] to the PR title. Once it is published, please do ping us again.

**Note:** Our approach to having `[hold pub]` labelled PRs may change in the near future once we have stricter rules on how long PRs can remain open.

**Comment by IAlibay** [comment]
On a more reviewer 3 type of comment - my general understanding is that the only way to prove that this approximation works is to compare it against one of the previously published methods (i.e. as defined in the Gabb et al. paper). In that case, I would ask that this validation method (which I assume is already written in a similar way?) also be added as an option for users to call & compare against.

**Comment by IAlibay** [comment]
@aakognole @orbeckst - any ideas on the status of this PR?

**Comment by orbeckst** [comment]
I think the primary question is if @aakognole published a paper where they applied their method.

**Comment by aakognole** [comment]
@orbeckst Thanks for your consideration. But this work was not published and I have moved on to another role. You may close the pull request.

**Comment by IAlibay** [comment]
Thanks for updating us @aakognole. Please do reopen the PR if the situation changes.

**Pull request closed without merging**
